File: src/main/java/javabot/operations/SeenOperation.java
package javabot.operations;

import java.text.DateFormat;
import java.util.ArrayList;
import java.util.List;

import javabot.BotEvent;
import javabot.Javabot;
import javabot.Message;
import javabot.dao.LogsDao;
import org.springframework.beans.factory.annotation.Autowired;
// Author : joed
// Date  : Apr 8, 2007

public class SeenOperation extends BotOperation {
    @Autowired
    private LogsDao dao;

    public SeenOperation(final Javabot javabot) {
        super(javabot);
    }

    @Override
    public List<Message> handleMessage(final BotEvent event) {
        final List<Message> messages = new ArrayList<Message>();
        final String message = event.getMessage().toLowerCase();
        final String channel = event.getChannel();
        final String sender = event.getSender();
        if (message.startsWith("seen ")) {
            final String key = message.substring("seen ".length());
            if (dao.isSeen(key, channel)) {
                messages.add(new Message(channel, event,
                    String.format("%s, %s was last seen at %s with the following entry: %s", sender, key,
                        DateFormat.getInstance().format(dao.getSeen(key, channel).getUpdated()),
                        dao.getSeen(key, channel).getMessage())));
                return messages;
            }
            messages.add(new Message(channel, event, sender + ", I have no information about \"" + key + "\""));
        }
        return messages;
    }
}

File: src/main/kotlin/javabot/operations/SeenOperation.kt
package javabot.operations

import com.antwerkz.sofia.Sofia
import jakarta.inject.Inject
import java.time.format.DateTimeFormatter
import javabot.Javabot
import javabot.Message
import javabot.dao.AdminDao
import javabot.dao.LogsDao

class SeenOperation @Inject constructor(bot: Javabot, adminDao: AdminDao, var dao: LogsDao) :
    BotOperation(bot, adminDao) {
    override fun handleMessage(event: Message): List<Message> {
        val responses = arrayListOf<Message>()
        val message = event.value
        val channel = event.channel
        if (
            channel != null &&
                "seen ".equals(message.substring(0, Math.min(message.length, 5)), ignoreCase = true)
        ) {
            val key = message.substring("seen ".length)
            val seen = dao.getSeen(channel.name, key)
            if (seen != null) {
                responses.add(
                    Message(
                        event,
                        Sofia.seenLast(
                            event.user.nick,
                            key,
                            seen.updated.format(FORMATTER),
                            seen.message
                        )
                    )
                )
            } else {
                responses.add(Message(event, Sofia.seenUnknown(event.user.nick, key)))
            }
        }
        return responses
    }

    companion object {
        val FORMATTER: DateTimeFormatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm")
    }
}
